packages/dom-adapters/src/BlockToolAdapter/index.ts

```diff
@@ -0,0 +1,196 @@
+  /**
+   * Handles beforeinput event from user input and updates model data
+   *
+   * We prevent beforeinput event of any type to handle it manually via model update
+   *
+   * @param event - beforeinput event
+   * @param input - input element
+   * @param key - data key input is attached to
+   */
+  #handleBeforeInputEvent(event: InputEvent, input: HTMLElement, key: DataKey): void {
+    /**
+     * We prevent all events to handle them manually via model update
+     */
+    event.preventDefault();
+
+    const inputType = event.inputType as InputType;
+    const targetRanges = event.getTargetRanges();
+    const range = targetRanges[0];
+
+    const start = getAbsoluteRangeOffset(input, range.startContainer, range.startOffset);
+    const end = getAbsoluteRangeOffset(input, range.endContainer, range.endOffset);
+
+    switch (inputType) {
+      case InputType.InsertReplacementText: {
+        this.#model.removeText(this.#blockIndex, key, start, end);
+
+        const data = event.dataTransfer!.getData('text/plain');
+
+        this.#model.insertText(this.#blockIndex, key, data, start);
+
+        break;
+      }
+      case InputType.InsertText:
+      /**
+       * @todo Handle composition events
+       */
+      case InputType.InsertCompositionText: {
+        /**
+         * If start and end aren't equal, it means that user selected some text and replaced it with new one
+         */
+        if (start !== end) {
+          this.#model.removeText(this.#blockIndex, key, start, end);
+        }
+
+        const data = event.data as string;
+
+        this.#model.insertText(this.#blockIndex, key, data, start);
+        break;
+      }
+
+      case InputType.DeleteContent:
+      case InputType.DeleteContentBackward:
+      case InputType.DeleteContentForward:
+      case InputType.DeleteByCut:
+      case InputType.DeleteByDrag:
+      case InputType.DeleteHardLineBackward:
+      case InputType.DeleteHardLineForward:
+      case InputType.DeleteSoftLineBackward:
+      case InputType.DeleteSoftLineForward:
+      case InputType.DeleteWordBackward:
+      case InputType.DeleteWordForward: {
+        this.#model.removeText(this.#blockIndex, key, start, end);
+
+        break;
+      }
+
+      default:
+    }
+  }
+
+  /**
+   * Handles model update events and updates DOM
+   *
+   * @param event - model update event
+   * @param input - attched input element
+   * @param key - data key input is attached to
+   * @param caretAdapter - caret adapter instance
+   */
+  #handleModelUpdate(event: ModelEvents, input: HTMLElement, key: DataKey, caretAdapter: CaretAdapter): void {
+    if (!(event instanceof TextAddedEvent) && !(event instanceof TextRemovedEvent)) {
+      return;
+    }
+
+    const [rangeIndex, dataIndex, blockIndex] = event.detail.index;
+
+    if (blockIndex !== this.#blockIndex) {
+      return;
+    }
+
+    if (dataIndex !== `data@${key}`) {
+      return;
+    }
+
+    if (!Array.isArray(rangeIndex)) {
+      return;
```

Comment:
it looks like exception

Reply:
this case still not clear. Add jsdoc please describing how it can happen

Reply:
Hm, I have removed it, not sure why it's still in the PR